File: server/src/routes/signup.rs
use actix_web::{web, HttpResponse, Responder};
use serde::Deserialize;
use dotenv::dotenv;
use std::env;
use unofficial_appwrite::client::{Client, ClientBuilder};
use unofficial_appwrite::error::Error;
use unofficial_appwrite::services::server::users::Users;
use unofficial_appwrite::id::ID;

#[derive(Deserialize)]
pub struct SignupData {
    email: String,
    username: String,
    password: String
}

fn build_client(endpoint: &str, id: &str, key: &str) -> Result<Client, Error> {
    ClientBuilder::default()
    .set_endpoint(endpoint)?
    .set_project(id)?
    .set_key(key)?
    .build()
}

pub async fn signup(data: web::Json<SignupData>) -> impl Responder {
    dotenv().ok();

    let api_endpoint = match env::var("APPWRITE_API_ENDPOINT") {
        Ok(endpoint) => endpoint,
        Err(_) => return HttpResponse::InternalServerError().body("APPWRITE_API_ENDPOINT was not defined in the .env file")
    };
    
    let project_id = match env::var("APPWRITE_PROJECT_ID") {
        Ok(id) => id,
        Err(_) => return HttpResponse::InternalServerError().body("APPWRITE_PROJECT_ID was not defined in the .env file")
    };

    let api_key = match env::var("APPWRITE_API_KEY") {
        Ok(endpoint) => endpoint,
        Err(_) => return HttpResponse::InternalServerError().body("APPWRITE_API_ENDPOINT was not defined in the .env file")
    };
    
    let client= match build_client(&api_endpoint, &project_id, &api_key) {
        Ok(c) => c,
        Err(_) => return HttpResponse::InternalServerError().body("Client failed to build")
    };
    
    let create_user = match Users::create(
        &client, 
        ID::unique(), 
        Some(&data.email),
        None, 
        Some(&data.password), 
        Some(&data.username)
    ).await {
        Ok(c) => c,
        Err(e) => return HttpResponse::InternalServerError().body(format!("{:?}", e))
    };
    
    HttpResponse::Ok().body(format!("{} has signed up successfully!", create_user.name))
}

File: server/routes/signup.go
package routes

import (
	"encoding/json"
	"server/utils"

	"github.com/appwrite/sdk-for-go/appwrite"
	"github.com/appwrite/sdk-for-go/id"
	"github.com/appwrite/sdk-for-go/query"

	"github.com/savsgio/atreugo/v11"

	"golang.org/x/crypto/bcrypt"
)

type SignupData struct {
	Email string `json:"email"`
	Username string `json:"username"`
	Password string `json:"password"`
}

func Signup(ctx *atreugo.RequestCtx) error {
	var signupRequest SignupData
	
	if err := json.Unmarshal(ctx.Request.Body(), &signupRequest); err != nil {
		return ctx.JSONResponse(map[string]interface{}{
			"successful" : false,
			"message": "Signup data was not provided correctly",
		}, 400)
	}
	
	client := utils.CreateClient()

	database := appwrite.NewDatabases(client)

	list, err := database.ListDocuments("mixtaper", "usernames", database.WithListDocumentsQueries(
		[]string{query.Equal("username", signupRequest.Username)},
	))

	if err != nil {
		return ctx.JSONResponse(map[string]interface{}{
			"successful" : false,
			"message": "Failed to fetch used usernames",
		}, 500)
	}

	if list.Total > 0 {
		return ctx.JSONResponse(map[string]interface{}{
			"successful" : false,
			"message": "Username is taken",
		}, 400)
	}
	
	hashed_password, err := bcrypt.GenerateFromPassword([]byte(signupRequest.Password), bcrypt.DefaultCost)
	
	if err != nil {
		return ctx.JSONResponse(map[string]interface{}{
			"successful" : false,
			"message": "Failed to hash password",
		}, 500)
	}
	
	users := appwrite.NewUsers(client)
	
	user, err := users.CreateBcryptUser(
		id.Unique(),
		signupRequest.Email,
		string(hashed_password),
		users.WithCreateBcryptUserName(signupRequest.Username),
	)
	
	if err != nil {
		return ctx.JSONResponse(map[string]interface{}{
			"successful" : false,
			"message": "Failed to create user",
		}, 500)
	}

	_, err = database.CreateDocument(
		"mixtaper",
		"usernames",
		user.Id,
		map[string]string{
			"username" : signupRequest.Username,
		},
	)

	if err != nil {
		return ctx.JSONResponse(map[string]interface{}{
			"successful" : false,
			"message": "Failed to add username to database",
		}, 500)
	}

	return ctx.JSONResponse(map[string]interface{}{
		"successful" : true,
		"message": user.Name + " has signed up successfully!",
	}, 200)
}